src/libraries/System.Net.Http/src/System/Net/Http/MetricsHandler.cs

```diff
@@ -0,0 +1,226 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http;
+
+internal sealed class MetricsHandler : HttpMessageHandlerStage
+{
+    private readonly HttpMessageHandler _innerHandler;
+    private readonly Meter _meter;
+    private readonly UpDownCounter<long> _currentRequests;
+    private readonly Counter<long> _failedRequests;
+    private readonly Histogram<double> _requestsDuration;
+
+    public MetricsHandler(HttpMessageHandler innerHandler, Meter meter)
+    {
+        _meter = meter;
+        _innerHandler = innerHandler;
+
+        _currentRequests = _meter.CreateUpDownCounter<long>(
+            "http-client-current-requests",
+            description: "Number of outbound HTTP requests that are currently active on the client.");
+        _failedRequests = _meter.CreateCounter<long>(
+            "http-client-failed-requests",
+            description: "Number of outbound HTTP requests that have failed.");
+        _requestsDuration = _meter.CreateHistogram<double>(
+            "http-client-request-duration",
+            unit: "s",
+            description: "The duration of outbound HTTP requests.");
+    }
+
+    public static Meter DefaultMeter { get; } = new SharedMeter();
+
+    internal override ValueTask<HttpResponseMessage> SendAsync(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
+    {
+        if (_currentRequests.Enabled || _failedRequests.Enabled || _requestsDuration.Enabled)
+        {
+            ArgumentNullException.ThrowIfNull(request);
+            return SendAsyncWithMetrics(request, async, cancellationToken);
+        }
+        else
+        {
+            return async ?
+                new ValueTask<HttpResponseMessage>(_innerHandler.SendAsync(request, cancellationToken)) :
+                new ValueTask<HttpResponseMessage>(_innerHandler.Send(request, cancellationToken));
+        }
+    }
+
+    private async ValueTask<HttpResponseMessage> SendAsyncWithMetrics(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
+    {
+        (long startTimestamp, bool recordCurrentRequests) = RequestStart(request);
+        HttpResponseMessage? response = null;
+        try
+        {
+            response = async ?
+                await _innerHandler.SendAsync(request, cancellationToken).ConfigureAwait(false) :
+                _innerHandler.Send(request, cancellationToken);
+            if (_failedRequests.Enabled)
+            {
+                // No exception has been thrown to the point of reading headers, but errors can still occur while buffering the response content in HttpClient.
+                // We need to report http-client-failed-requests if it happens.
+                response.RequestFailedMetricsLogger = LogRequestFailed;
```

Comment:
This violates handler layering by providing access to a method from `MetricsHandler` to `HttpClient`. Alternative solutions:
- Wrap content and stream, so we can detect errors when reading the streams. This would be stronger than what we have with counters and events today, reporting `http-client-failed-requests` when user code is reading the content stream, but would come with an allocation + complexity cost.
- Do not report `http-client-failed-requests` for errors happening during content buffering, which would be weaker than what we have with counters and events, and might be misleading for users.